src/App.tsx

```diff
@@ -19,10 +39,25 @@ class App extends React.Component {
           >
             Learn React
           </a>
+          <p>Toggle this on and off to test out Redux:</p>
+          <p>
+            <button
+              onClick={this.handleToggleClick}
+              style={{ padding: '32px' }}
+            >
+              {toggledOn ? 'OFF' : 'ON'}
+            </button>
+          </p>
         </header>
       </div>
     );
   }
 }
 
-export default App;
+const mapStateToProps = (state: ApplicationState) => {
+  return {
+    toggledOn: state.example.toggledOn,
+  };
+};
+
+export default connect(mapStateToProps)(App);
```

Comment:
Hmm, don't we need to do something like:

```
const App: React.ComponentType<PublicProps> = connect(mapStateToProps)(AppBase);
export default App;
```

In order to ensure type safety when using this component? Even if TS is smart enough to understand which props come from state (and I'm not sure how it can do that - I guess by magic), we still have the `ConnectedReduxProps` in the props listed for the default component.

Reply:
We do not! TypeScript figures everything out. Here are some examples:

Passing an unknown prop named `foo` to a component:

```
Type error: Type '{ foo: string; }' is not assignable to type 'IntrinsicAttributes & IntrinsicClassAttributes<Component<Pick<PropsFromState & ConnectedReduxProps<AnyAction>, never>, any, any>> & Readonly<{ children?: ReactNode; }> & Readonly<...>'.
  Property 'foo' does not exist on type 'IntrinsicAttributes & IntrinsicClassAttributes<Component<Pick<PropsFromState & ConnectedReduxProps<AnyAction>, never>, any, any>> & Readonly<{ children?: ReactNode; }> & Readonly<...>'.  TS2322

    13 |   <Provider store={store}>
    14 |     <BrowserRouter>
  > 15 |       <App foo="bar" />
       |        ^
    16 |     </BrowserRouter>
    17 |   </Provider>,
    18 |   document.getElementById('root') as HTMLElement,
```

Passing an  *already mapped* prop to a component:

```
Type error: Type '{ toggledOn: boolean; }' is not assignable to type 'IntrinsicAttributes & IntrinsicClassAttributes<Component<Pick<PropsFromState & ConnectedReduxProps<AnyAction>, never>, any, any>> & Readonly<{ children?: ReactNode; }> & Readonly<...>'.
  Property 'toggledOn' does not exist on type 'IntrinsicAttributes & IntrinsicClassAttributes<Component<Pick<PropsFromState & ConnectedReduxProps<AnyAction>, never>, any, any>> & Readonly<{ children?: ReactNode; }> & Readonly<...>'.  TS2322

    13 |   <Provider store={store}>
    14 |     <BrowserRouter>
  > 15 |       <App toggledOn={true} />
       |        ^
    16 |     </BrowserRouter>
    17 |   </Provider>,
    18 |   document.getElementById('root') as HTMLElement,
```

Passing a special Redux prop (like `dispatch`):

```
Type error: Type '{ dispatch: () => void; }' is not assignable to type 'IntrinsicAttributes & IntrinsicClassAttributes<Component<Pick<PropsFromState & ConnectedReduxProps<AnyAction>, never>, any, any>> & Readonly<{ children?: ReactNode; }> & Readonly<...>'.
  Property 'dispatch' does not exist on type 'IntrinsicAttributes & IntrinsicClassAttributes<Component<Pick<PropsFromState & ConnectedReduxProps<AnyAction>, never>, any, any>> & Readonly<{ children?: ReactNode; }> & Readonly<...>'.  TS2322

    13 |   <Provider store={store}>
    14 |     <BrowserRouter>
  > 15 |       <App dispatch={() => {}} />
       |        ^
    16 |     </BrowserRouter>
    17 |   </Provider>,
    18 |   document.getElementById('root') as HTMLElement,
```

I don't know how it figured all of this out but my gut instinct is that the type definitions imported from `react-redux` have been implemented to do all the right things. This gives me a warm fuzzy feeling that TypeScript "just works" a lot better than Flow.

Reply:
>  This gives me a warm fuzzy feeling that TypeScript "just works" a lot better than Flow.

💯